Update pod security standards to use PodOS field (#35772)
kubernetes/website: content/en/docs/concepts/security/pod-security-standards.md

		<tr>
			<td style="white-space: nowrap">Privilege Escalation (v1.8+)</td>
			<td>

> 				<p>On Linux, do not allow privilege escalation (such as via set-user-ID or set-group-ID file mode). This control only applies when you set `spec.os` to <code>linux</code>, or if you do not specify a Pod OS. Versions of Kubernetes before v1.25 enforce this control regardless of operating system.</p>

Reviewer: Rather than integrating this text in here, I'd suggest just having a standard header that you can add to all the linux-only fields. I prefer this approach because it makes it easier to identify the linux-only fields with a visual scan.

Maybe something like this:

```
<p><em><a href="#what-profiles-should-i-apply-to-my-windows-pods">Linux only</a> in v1.25+ (`spec.os != windows`)</em></p>
```

Reviewer 2: It's a bit odd to hyperlink “Linux only” to `#what-profiles-should-i-apply-to-my-windows-pods` - maybe we could have a new page section that covers OS specifics, with a subheading for Linux and another subheading for Windows.

		<tr>
  			<td style="white-space: nowrap">Seccomp (v1.19+)</td>
  			<td>

>   				<p>Seccomp profile must be explicitly set to one of the allowed values. Both the <code>Unconfined</code> profile and the <em>absence</em> of a profile are prohibited. <em><a href="#policies-specific-to-linux">This is Linux only policy</a> in v1.25+ <code>(spec.os.name != windows)</code></em></p>

Reviewer 2: Is this also right?
```suggestion
		<p>Seccomp profile must be explicitly set to one of the allowed values. Both the <code>Unconfined</code> profile and the <em>absence</em> of a profile are prohibited. <em><a href="#policies-specific-to-linux">This is Linux only policy</a> in v1.25+ <code>(.spec.os.name != "windows")</code></em></p>
```

Reviewer 2: It might be safer to use `{{ < skew currentVersion >}}` rather than `v1.25+`, because we are not able to foretell what logic all future minor releases of Kubernetes v1.x might use here.

>   				<p>On Linux, require that the seccomp profile is explicitly set to one of the allowed values. Both the <code>Unconfined</code> profile and the <em>absence</em> of a profile are prohibited.  This control only applies when you set `spec.os` to <code>linux</code>, or if you do not specify a Pod OS. Versions of Kubernetes before v1.25 enforce this control regardless of operating system.</p>

Reviewer: > ... when you set `spec.os` to `linux` ...

It's actually coded the other way - skip the check when it's set to `windows`. A subtle difference, but it could matter if we added another OS.

Reviewer 2: I wonder if I should also file a feature request issue to invert the logic there?

Reviewer 3: exempting only for an explicitly windows pod is intentional, so we default in the safe direction for new hitherto unknown OSes

>   				<p>On Linux, require that the seccomp profile is explicitly set to one of the allowed values. Both the <code>Unconfined</code> profile and the <em>absence</em> of a profile are prohibited.  This control only applies when you set `spec.os` to <code>linux</code>, or if you do not specify a Pod OS. Versions of Kubernetes before v1.25 enforce this control regardless of operating system.</p>

Reviewer: Unfortunately markdown syntax isn't processed in this table, so you need to use raw HTML tags.

```suggestion
  				<p>On Linux, require that the seccomp profile is explicitly set to one of the allowed values. Both the <code>Unconfined</code> profile and the <em>absence</em> of a profile are prohibited.  This control only applies when you set <code>spec.os</code> to <code>linux</code>, or if you do not specify a Pod OS. Versions of Kubernetes before v1.25 enforce this control regardless of operating system.</p>
```

Reviewer 2: Also, is it actually `.spec.os.name`?

			<td>
				<p>
					Containers must drop <code>ALL</code> capabilities, and are only permitted to add back

> 					the <code>NET_BIND_SERVICE</code> capability. Starting v1.25, windows pods would be exempted from this control if pod.spec.os.Name is set to <code>windows</code>

Reviewer 4: Other k8s docs capitalize the "W" in "Windows pods". I suggest changing the lowercase "windows pods" instances in this document.

Author: Sure. Will change it.

> 					the <code>NET_BIND_SERVICE</code> capability. <em><a href="#policies-specific-to-linux">This is Linux only policy</a> in v1.25+ <code>(spec.os.name != windows)</code></em>

Reviewer 2: Is this also right?
```suggestion
 					the <code>NET_BIND_SERVICE</code> capability. <em><a href="#policies-specific-to-linux">This is Linux only policy</a> in v1.25+ <code>(.spec.os.name != "windows")</code></em>
```

- [Kyverno](https://kyverno.io/policies/pod-security/)
- [OPA Gatekeeper](https://github.com/open-policy-agent/gatekeeper)

## Pod OS field

> Windows in Kubernetes has some limitations and differentiators from standard Linux-based

Reviewer 2: ```suggestion
Kubernetes lets you use nodes that run either Linux or Windows. You can mix both kinds of
node in one cluster.
Windows in Kubernetes has some limitations and differentiators from Linux-based
```

> workloads. Specifically, many of the Pod SecurityContext fields

Reviewer 2: ```suggestion
workloads. Specifically, many of the Pod `securityContext` fields
```

[have no effect on Windows](/docs/concepts/windows/intro/#compatibility-v1-pod-spec-containers-securitycontext).

> 

Reviewer: Add a warning here about how Kubelets prior to v1.24 don't enforce the pod OS field, and if a cluster has nodes on versions earlier than v1.24 the restricted policies should be pinned to a version prior to v1.25.

Author: Added note here:  https://github.com/kubernetes/website/pull/35772/files#diff-1eb43daa07b175394d3d428f40b35e9010ad719ce9fa8c0e858a0f773e4330faR472

> ### Pod API validation changes

Reviewer 2: ```suggestion
### Pod security checks based on Pod operating system
```

Document the behavior, _not_ the change in behavior (that's for the release notes)/

For Kubernetes v{{< skew currentVersion >}} it is good practice to specify
the `.spec.os` field for every Pod. The `.spec.os` field is available in Kubernetes
v1.23 and later but was behind a
[feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
for Kubernetes v1.23 and v1.24.
If you set the `pod.spec.os.name` to `linux`, the enforced security controls should be pertinent to Linux.
Setting the `pod.spec.os.name` field to `windows` means that you don't need to apply Linux-specific restrictions 
and the API validation for the pod object would fail if you apply Linux-specific restrictions to such pods.
For example, a Windows Pod does not need to explicitly drop all capabilities for each container, even if 
you are applying the `restricted` Pod security standard. If you don't specify the `pod.spec.os.name`, 
Kubernetes takes a precautionary approach and applies all the Pod security measures that could be relevant 
to Linux. 

### Restricted Pod Security Standard changes

> Another important change starting Kubernetes v{{ < skew currentVersion >}} is that the pod restricted standard

Reviewer 2: ```suggestion
Another important change, made in Kubernetes v1.25, is that the _restricted_ Pod security standard
```
If you can, rewrite the text to define and document the current (oost-v1.25) behaviour, rather than the state change.

Author: @tallclair - You mean a link to the changelog on this page?

has been updated to use the `pod.spec.os.name` field. Based on the OS name, certain policies that are specific

> to a particualr OS can be relaxed for the other OS. 

Reviewer 2: (nit)
```suggestion
to a particular OS can be relaxed for the other OS.
```

#### Policies specific to Linux
- Privilege Escalation
- Seccomp
- Linux Capabilities

> 

Reviewer 2: ```suggestion

#### Policies specific to Windows

```
?

Author: We don't have any policies specific to Windows.

Author: This is not needed.

> #### Policies specific to Linux
> - Privilege Escalation
> - Seccomp
> - Linux Capabilities
> 

Reviewer: ```suggestion
#### OS-specific policy controls
Restrictions on the following controls are only required if `.spec.os.name` is not `windows`:
- Privilege Escalation
- Seccomp
- Linux Capabilities
```

The use of the HostProcess flag to create a HostProcess pod should only be done in alignment with the privileged policy.
Creation of a Windows HostProcess pod is blocked under the baseline and restricted policies,

> so any HostProcess pod should be considered privileged.

Reviewer: I think we can drop this. It's already covered in the policy table, and doesn't behave any differently with pod os set.

Windows in Kubernetes has some limitations and differentiators from standard Linux-based
workloads. Specifically, many of the Pod SecurityContext fields
[have no effect on Windows](/docs/concepts/windows/intro/#compatibility-v1-pod-spec-containers-securitycontext).

> In v1.25, the `pod.spec.os` field has graduated to stable which helps identify the windows pods authoritatively. Pod Security admission plugin has been updated to use this field which allows certain relaxations for the windows pods. The restricted profile has been updated to ensure that the Linux-specific restrictions such as seccomp profile, disallowing privilege escalation, capabilities are ignored if the `pod.spec.os.Name` is set to `windows`.

Author: I am fine with @tallclair any concerns?

> In v1.25, the `pod.spec.os` field has graduated to stable which helps identify the windows pods authoritatively. Pod Security admission plugin has been updated to use this field which allows certain relaxations for the windows pods. The restricted profile has been updated to ensure that the Linux-specific restrictions such as seccomp profile, disallowing privilege escalation, capabilities are ignored if the `pod.spec.os.Name` is set to `windows`.

Reviewer: I agree. I was going to suggest the same thing.
